### ISSUE TITLE
Minimize bounds check failure

### ISSUE THREAD

**Issue by bashtage** [scipy/scipy#11367]
The bounds check in `minimize`, https://github.com/scipy/scipy/blame/master/scipy/optimize/_numdiff.py#L392, seems to be raising for valid input.

This is showing up using pre-release NumPy 1.19 and SciPy 1.5.

We have a test run failure here: 

https://travis-ci.org/statsmodels/statsmodels/jobs/637413106

Pytest reports the parameter values:

```
x0 = array([ 0.45676527, -0.        ,  0.51941534, 14.20525598,  0.12234215,
       -1.4559698 , -5.49189273, -3.87943515, -2.97319091])
method = '2-point', rel_step = None, abs_step = 1e-08, f0 = array([1.97457067])
bounds = (array([  0.,   0.,   0.,   0.,   0., -inf, -inf, -inf, -inf]), array([ 1.,  1.,  1., inf, inf, inf, inf, inf, inf]))
```

`x0` is in bounds which requires the first 5 elements to be >=0 and the first 3 to be <=1. The only notable value is x0[1] which is -0., which should still be OK with the test

```
        if np.any((x0 < lb) | (x0 > ub)):
           raise ValueError("`x0` violates bound constraints.")
```

In the statsmodels code I set each parameter to be at the boundary if outside before calling minimize.

#### Reproducing code example:

I can't easily repro but this test fails each time. 

#### Error message:
```
statsmodels/tsa/tests/test_holtwinters.py:329: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
statsmodels/tsa/holtwinters.py:728: in fit
    res = minimize(func, p[xi], args=args, bounds=_bounds)
venv/lib/python3.7/site-packages/scipy/optimize/_minimize.py:611: in minimize
    callback=callback, **options)
venv/lib/python3.7/site-packages/scipy/optimize/lbfgsb.py:357: in _minimize_lbfgsb
    f, g = func_and_grad(x)
venv/lib/python3.7/site-packages/scipy/optimize/_differentiable_functions.py:202: in fun_and_grad
    self._update_grad()
venv/lib/python3.7/site-packages/scipy/optimize/_differentiable_functions.py:172: in _update_grad
    self._update_grad_impl()
venv/lib/python3.7/site-packages/scipy/optimize/_differentiable_functions.py:93: in update_grad
    **finite_diff_options)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
fun = <function ScalarFunction.__init__.<locals>.fun_wrapped at 0x7f18e28e80d0>
x0 = array([ 0.45676527, -0.        ,  0.51941534, 14.20525598,  0.12234215,
       -1.4559698 , -5.49189273, -3.87943515, -2.97319091])
method = '2-point', rel_step = None, abs_step = 1e-08, f0 = array([1.97457067])
bounds = (array([  0.,   0.,   0.,   0.,   0., -inf, -inf, -inf, -inf]), array([ 1.,  1.,  1., inf, inf, inf, inf, inf, inf]))
sparsity = None, as_linear_operator = False, args = (), kwargs = {}
    def approx_derivative(fun, x0, method='3-point', rel_step=None, abs_step=None,
                          f0=None, bounds=(-np.inf, np.inf), sparsity=None,
                          as_linear_operator=False, args=(), kwargs={}):
        """Compute finite difference approximation of the derivatives of a
        vector-valued function.
    
        If a function maps from R^n to R^m, its derivatives form m-by-n matrix
        called the Jacobian, where an element (i, j) is a partial derivative of
        f[i] with respect to x[j].
    
        Parameters
        ----------
        fun : callable
            Function of which to estimate the derivatives. The argument x
            passed to this function is ndarray of shape (n,) (never a scalar
            even if n=1). It must return 1-D array_like of shape (m,) or a scalar.
        x0 : array_like of shape (n,) or float
            Point at which to estimate the derivatives. Float will be converted
            to a 1-D array.
        method : {'3-point', '2-point', 'cs'}, optional
            Finite difference method to use:
                - '2-point' - use the first order accuracy forward or backward
                              difference.
                - '3-point' - use central difference in interior points and the
                              second order accuracy forward or backward difference
                              near the boundary.
                - 'cs' - use a complex-step finite difference scheme. This assumes
                         that the user function is real-valued and can be
                         analytically continued to the complex plane. Otherwise,
                         produces bogus results.
        rel_step : None or array_like, optional
            Relative step size to use. The absolute step size is computed as
            ``h = rel_step * sign(x0) * max(1, abs(x0))``, possibly adjusted to
            fit into the bounds. For ``method='3-point'`` the sign of `h` is
            ignored. If None (default) then step is selected automatically,
            see Notes.
        abs_step : array_like, optional
            Absolute step size to use, possibly adjusted to fit into the bounds.
            For ``method='3-point'`` the sign of `abs_step` is ignored. By default
            relative steps are used, only if ``abs_step is not None`` are absolute
            steps used.
        f0 : None or array_like, optional
            If not None it is assumed to be equal to ``fun(x0)``, in  this case
            the ``fun(x0)`` is not called. Default is None.
        bounds : tuple of array_like, optional
            Lower and upper bounds on independent variables. Defaults to no bounds.
            Each bound must match the size of `x0` or be a scalar, in the latter
            case the bound will be the same for all variables. Use it to limit the
            range of function evaluation. Bounds checking is not implemented
            when `as_linear_operator` is True.
        sparsity : {None, array_like, sparse matrix, 2-tuple}, optional
            Defines a sparsity structure of the Jacobian matrix. If the Jacobian
            matrix is known to have only few non-zero elements in each row, then
            it's possible to estimate its several columns by a single function
            evaluation [3]_. To perform such economic computations two ingredients
            are required:
    
            * structure : array_like or sparse matrix of shape (m, n). A zero
              element means that a corresponding element of the Jacobian
              identically equals to zero.
            * groups : array_like of shape (n,). A column grouping for a given
              sparsity structure, use `group_columns` to obtain it.
    
            A single array or a sparse matrix is interpreted as a sparsity
            structure, and groups are computed inside the function. A tuple is
            interpreted as (structure, groups). If None (default), a standard
            dense differencing will be used.
    
            Note, that sparse differencing makes sense only for large Jacobian
            matrices where each row contains few non-zero elements.
        as_linear_operator : bool, optional
            When True the function returns an `scipy.sparse.linalg.LinearOperator`.
            Otherwise it returns a dense array or a sparse matrix depending on
            `sparsity`. The linear operator provides an efficient way of computing
            ``J.dot(p)`` for any vector ``p`` of shape (n,), but does not allow
            direct access to individual elements of the matrix. By default
            `as_linear_operator` is False.
        args, kwargs : tuple and dict, optional
            Additional arguments passed to `fun`. Both empty by default.
            The calling signature is ``fun(x, *args, **kwargs)``.
    
        Returns
        -------
        J : {ndarray, sparse matrix, LinearOperator}
            Finite difference approximation of the Jacobian matrix.
            If `as_linear_operator` is True returns a LinearOperator
            with shape (m, n). Otherwise it returns a dense array or sparse
            matrix depending on how `sparsity` is defined. If `sparsity`
            is None then a ndarray with shape (m, n) is returned. If
            `sparsity` is not None returns a csr_matrix with shape (m, n).
            For sparse matrices and linear operators it is always returned as
            a 2-D structure, for ndarrays, if m=1 it is returned
            as a 1-D gradient array with shape (n,).
    
        See Also
        --------
        check_derivative : Check correctness of a function computing derivatives.
    
        Notes
        -----
        If `rel_step` is not provided, it assigned to ``EPS**(1/s)``, where EPS is
        machine epsilon for float64 numbers, s=2 for '2-point' method and s=3 for
        '3-point' method. Such relative step approximately minimizes a sum of
        truncation and round-off errors, see [1]_. Relative steps are used by
        default. However, absolute steps are used when ``abs_step is not None``.
        If any of the absolute steps produces an indistinguishable difference from
        the original `x0`, ``(x0 + abs_step) - x0 == 0``, then a relative step is
        substituted for that particular entry.
    
        A finite difference scheme for '3-point' method is selected automatically.
        The well-known central difference scheme is used for points sufficiently
        far from the boundary, and 3-point forward or backward scheme is used for
        points near the boundary. Both schemes have the second-order accuracy in
        terms of Taylor expansion. Refer to [2]_ for the formulas of 3-point
        forward and backward difference schemes.
    
        For dense differencing when m=1 Jacobian is returned with a shape (n,),
        on the other hand when n=1 Jacobian is returned with a shape (m, 1).
        Our motivation is the following: a) It handles a case of gradient
        computation (m=1) in a conventional way. b) It clearly separates these two
        different cases. b) In all cases np.atleast_2d can be called to get 2-D
        Jacobian with correct dimensions.
    
        References
        ----------
        .. [1] W. H. Press et. al. "Numerical Recipes. The Art of Scientific
               Computing. 3rd edition", sec. 5.7.
    
        .. [2] A. Curtis, M. J. D. Powell, and J. Reid, "On the estimation of
               sparse Jacobian matrices", Journal of the Institute of Mathematics
               and its Applications, 13 (1974), pp. 117-120.
    
        .. [3] B. Fornberg, "Generation of Finite Difference Formulas on
               Arbitrarily Spaced Grids", Mathematics of Computation 51, 1988.
    
        Examples
        --------
        >>> import numpy as np
        >>> from scipy.optimize import approx_derivative
        >>>
        >>> def f(x, c1, c2):
        ...     return np.array([x[0] * np.sin(c1 * x[1]),
        ...                      x[0] * np.cos(c2 * x[1])])
        ...
        >>> x0 = np.array([1.0, 0.5 * np.pi])
        >>> approx_derivative(f, x0, args=(1, 2))
        array([[ 1.,  0.],
               [-1.,  0.]])
    
        Bounds can be used to limit the region of function evaluation.
        In the example below we compute left and right derivative at point 1.0.
    
        >>> def g(x):
        ...     return x**2 if x >= 1 else x
        ...
        >>> x0 = 1.0
        >>> approx_derivative(g, x0, bounds=(-np.inf, 1.0))
        array([ 1.])
        >>> approx_derivative(g, x0, bounds=(1.0, np.inf))
        array([ 2.])
        """
        if method not in ['2-point', '3-point', 'cs']:
            raise ValueError("Unknown method '%s'. " % method)
    
        x0 = np.atleast_1d(x0)
        if x0.ndim > 1:
            raise ValueError("`x0` must have at most 1 dimension.")
    
        lb, ub = _prepare_bounds(bounds, x0)
    
        if lb.shape != x0.shape or ub.shape != x0.shape:
            raise ValueError("Inconsistent shapes between bounds and `x0`.")
    
        if as_linear_operator and not (np.all(np.isinf(lb))
                                       and np.all(np.isinf(ub))):
            raise ValueError("Bounds not supported when "
                             "`as_linear_operator` is True.")
    
        def fun_wrapped(x):
            f = np.atleast_1d(fun(x, *args, **kwargs))
            if f.ndim > 1:
                raise RuntimeError("`fun` return value has "
                                   "more than 1 dimension.")
            return f
    
        if f0 is None:
            f0 = fun_wrapped(x0)
        else:
            f0 = np.atleast_1d(f0)
            if f0.ndim > 1:
                raise ValueError("`f0` passed has more than 1 dimension.")
    
        if np.any((x0 < lb) | (x0 > ub)):
>           raise ValueError("`x0` violates bound constraints.")
E           ValueError: `x0` violates bound constraints.```
```

#### Scipy/Numpy/Python version information:
```
Python: 3.7.1.final.0
OS: Linux 4.15.0-1028-gcp #29~16.04.1-Ubuntu SMP Tue Feb 12 16:31:10 UTC 2019 x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
statsmodels: v0.11.0rc2 
cython: 0.29.14 
numpy: 1.19.0.dev0+674d96e 
scipy: 1.5.0.dev0+06fa3e0 
pandas: 1.0.0rc0+32.g13b22fd94 
    dateutil: 2.8.1 
patsy: 0.5.1 
```

**Comment by andyfaff** [comment]
I obtained statsmodels at https://github.com/statsmodels/statsmodels/commit/3e72ebb8875ab0e3f0c62ecb7873042bb3d71083 and tried to run the statsmodel tests on my own computer (macOS).

Python: 3.7.6
numpy (pip): 1.18.1
scipy: 1.5.0.dev0+a849b8c (tip of scipy master branch)

The statsmodels test suite started crashing in `statsmodels/graphics/tests/test_functional.py`. However, the entire holtwinters test series seems to have run successfully:
```
platform darwin -- Python 3.7.6, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
Using --randomly-seed=1579105345
rootdir: /Users/andrew/Documents/Andy/programming/statsmodels, inifile: setup.cfg, testpaths: statsmodels
plugins: randomly-3.2.1, xdist-1.31.0, forked-1.1.3
collected 13028 items                                                                                                                                                                                                                                                     

statsmodels/tsa/arima/estimators/tests/test_statespace.py ..                                                                                                                                                                                                        [  0%]
statsmodels/tsa/arima/tests/test_tools.py ..........                                                                                                                                                                                                                [  0%]
statsmodels/tsa/tests/test_holtwinters.py ...................................................X.   
```

From the statsmodels travis test output the issue arises at [line 357](https://github.com/scipy/scipy/blob/master/scipy/optimize/lbfgsb.py#L357) after all initialisation has been done, and after at least one round of `_lbfgsb.setulb`.

My initial suspicion is that the issue was perhaps related to https://github.com/scipy/scipy/pull/10673, which now uses different code for numerical differentiation, `optimize._numdiff`. This code should respect bounds when a finite diff is carried out against a parameter limit. 
However, it's possible that it's also related to https://github.com/scipy/scipy/pull/10942, where the LBFGSB code proposes a step that is slightly outside the bounds. This would cause the `func_and_grad` evaluation to complain.

The stack trace reported by @bashtage is bemusing though. The Exception raised is because `x0 violates bound constraints`. The value of `x0` and `bounds` reported by travis are:
```
x0 = array([ 0.45676527, -0.        ,  0.51941534, 14.20525598,  0.12234215,
       -1.4559698 , -5.49189273, -3.87943515, -2.97319091])
bounds = (array([  0.,   0.,   0.,   0.,   0., -inf, -inf, -inf, -inf]), array([ 1.,  1.,  1., inf, inf, inf, inf, inf, inf]))
```
If I do the test:
```
from scipy.optimize._numdiff import _prepare_bounds
lb, ub = _prepare_bounds(bounds, x0)
assert not np.any((x0 < lb) | (x0 > ub))
```
then there are no issues on my machine. I wonder if the `x0` array reported by pytest is printing out what we need here, are there array printing options in numpy that would cause a number less than zero to be printed out as `-0.` (`x0[0]`)? We need full precision here.

@bashtage, a helpful way of debugging this further (because I can't reproduce the error on my machine) would be to insert some debugging code inside the relevant function being supplied to `minimize`. i.e.

```
def myfun(x):
    lb = array([  0.,   0.,   0.,   0.,   0., -inf, -inf, -inf, -inf])
    ub = array([ 1.,  1.,  1., inf, inf, inf, inf, inf, inf])
    try:
        assert not np.any((x < lb) | (x > ub))
     except AssertionError as e:
         print(x)
         print(x < lb)
         print(x > ub)
         raise e
```

**Comment by bashtage** [comment]
Thanks. I think the precision may be limited and it could be 1e-18 or so.  I'll have to try and do some (frustrating) travis debugging since I can only reproduce on travis using pre-release packages..

**Comment by andyfaff** [comment]
If it turns out that a value is just slightly outside the bounds then #10942 might provide the fix, but I'm not convinced about the tests there.

Perhaps the test debugging could be done on a local machine with Docker. You can use https://github.com/scipy/scipy/blob/master/tools/Dockerfile to setup an environment for scipy development. I don't know how you go about installing bleeding edge numpy though.

**Comment by andyfaff** [comment]
I can reproduce this using a Docker container with scipy@master and statsmodels@master, with Python 3.7 and numpy 1.18.1.

The relevant pytest output is:
```
fun = <function ScalarFunction.__init__.<locals>.fun_wrapped at 0x7f3edfeaeb70>, x0 = array([ 4.56765267e-01, -3.46944695e-18,  5.19415339e-01,  1.42052560e+01,
        1.22342148e-01, -1.45596980e+00, -5.49189273e+00, -3.87943515e+00,
       -2.97319091e+00])
method = '2-point', rel_step = None, abs_step = 1e-08, f0 = array([1.97457067]), bounds = (array([  0.,   0.,   0.,   0.,   0., -inf, -inf, -inf, -inf]), array([ 1.,  1.,  1., inf, inf, inf, inf, inf, inf])), sparsity = None, as_linear_operator = False, args = ()
kwargs = {}
```
What's relevant here is `x0`:
```
 x0 = array([ 4.56765267e-01, -3.46944695e-18,  5.19415339e-01,  1.42052560e+01,
        1.22342148e-01, -1.45596980e+00, -5.49189273e+00, -3.87943515e+00,
       -2.97319091e+00])
```
i.e. the `x0[1]` provided to `func_and_grad` is below the lower bounds of 0, which is why `optimize._numdiff.approx_derivative` raises the ValueError.  It's only a very little outside the bounds, so I suspect that the problem will be fixed by #10942.

**Comment by andyfaff** [comment]
With the changes from #10942 applied the statsmodel test now passes in my Docker setup. Thanks for reporting this @bashtage.

**Comment by bashtage** [comment]
It seems that it is getting some very small negative values with a 0 lower constraint:

```
fun = <function ScalarFunction.__init__.<locals>.fun_wrapped at 0x7fb2554bbdd0>
x0 = array([ 4.56765267e-01, -3.46944695e-18,  5.19415339e-01,  1.42052560e+01,
        1.22342148e-01, -1.45596980e+00, -5.49189273e+00, -3.87943515e+00,
       -2.97319091e+00])
method = '2-point', rel_step = None, abs_step = 1e-08, f0 = array([1.97457067])
bounds = (array([  0.,   0.,   0.,   0.,   0., -inf, -inf, -inf, -inf]), array([ 1.,  1.,  1., inf, inf, inf, inf, inf, inf]))
sparsity = None, as_linear_operator = False, args = (), kwargs = {}
```
Note `x[1]` which is `> -np.finfo(double).eps` but `< 0`.

**Comment by bashtage** [comment]
Thanks @andyfaff .

**Comment by andyfaff** [comment]
@bashtage this should be closed by #10942. If it's not, then please let me know and we can reopen.
